v2-roadmap.md

```diff
@@ -0,0 +1,201 @@
+
+* Pydantic implementation is calling the HTTP implementation for this, adding unnecessary overhead
+* the Kafka binding uses its own implementation, creating code duplication and unnecessary maintenance efforts
+
+We should have a single `CloudEvent` class used for all bindings/formats since the event
+is the Python representation of the event data, which is the same no matter what binding
+is used. (we potentially could receive a HTTP event and forward it as it is using Kafka).
+
+The specific bindings and formats implementations should be separate from the `CloudEvent`
+class.
+
+This proposal contains some base rules for the package architecture and how we could implement
+the `CloudEvent` spec in this package.
+
+## Requirements
+
+In order to be able to transform a python CloudEvent to/from a specific
+protocol binding, this SDK has the following responsibilities:
+
+* Identifying what attributes have to be marshalled/unmarshalled (identifying the cloudevents extensions)
+* Validate required fields and generation of defaults
+* Marshalling/unmarshalling the cloudevents core and extensions fields
+* Marshalling/unmarshalling the `data` field
+* Encode/decode the marshalled fields to/from the specific binding format
+
+## Modules
+
+We should separate the implementation of the requirements so that they can be tested independently and
+each component be used for composition.
+
+Ideally we should have these modules (naming TBC):
+
+* `cloudevents.event` - This would contain the `CloudEvent` base class. This class is the same for all bindings/formats.
+* `cloudevents.formats` - This would contain the structured formats implementations (JSON, AVRO, Protobuf)
+* `cloudevents.bindings` - This would contain the specific binding implementations (kafka, http)
+
+This matches the naming and concepts we have in the `CloudEvent` spec
+
+## Matching requirements and modules
+
+### The `cloudevents.event` module and `CloudEvent` class
+
+The `CloudEvent` class would satisfy the following requirements:
+
+* Validate required fields and generation of default values
+* Marshalling/unmarshalling the `data` field
+* Identifying what attributes have to be marshalled/unmarshalled
+
+The `data` field depends on the specific event and the `contentdatatype` attribute and
+its marshalling/unmarshalling logic should be the same for any binding/format. It seems
+a `CloudEvent` class responsibility to take care of this. (We should implement this logic
+in overridable methods)
+
+Ideally the end user will be able to extend the `CloudEvent` class for their necessity. Ie:
+
+* adding new extensions fields
+* adding fields/methods not meant to be marshalled (internal fields)
+* implementing new defaults (ie. creating a `MyEvent` class with `type="myevent"` default)
+
+We also want a DTO that can be accepted by the most common libraries (ie. JSON) to keep the
+`CloudEvent` class decoupled from the formats/bindings. Using a dictionary seems a good idea.
```

Comment:
I would expect the end user to never deal with this DTO. They should always rely on either the `CloudEvent` class or the specific format output, which could be a class (ie. `KafkaEvent`)

This DTO structure was thought to be used only internally, to move data between the base class, the formats and the bindings.

I thought about dictionaries because they are fast, and libraries we might want to use for formats and bindings will probably support them out of the box (the classic example would be the [JSON library](https://docs.python.org/3/library/json.html#json-to-py-table) that maps JSON objects to dicts)

Reply:
We could also keep the existing implementation interface where we have "get_attributes" and "get_data" and use them in bindings and formats.

If we find at some point that having a dict in the middle might help we can introduce it.